File: AndroidArchitecture/mvp/src/main/java/com/codearms/maoqiqi/app/data/source/remote/TasksRemoteDataSource.java
package com.codearms.maoqiqi.app.data.source.remote;

import android.os.Handler;

import com.codearms.maoqiqi.app.data.TaskBean;
import com.codearms.maoqiqi.app.data.source.TasksDataSource;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Access the server API interface data as a data source.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 15:42
 */
public class TasksRemoteDataSource implements TasksDataSource {

    private static final int SERVICE_LATENCY_IN_MILLIS = 5000;

    private static TasksRemoteDataSource INSTANCE;

    private static final Map<String, TaskBean> TASKS_SERVICE_DATA;

    static {
        TASKS_SERVICE_DATA = new LinkedHashMap<>(2);
        addTask("Build tower in Pisa", "Ground looks good, no foundation work required.");
        addTask("Finish bridge in Tacoma", "Found awesome girders at half the cost!");
    }

    private TasksRemoteDataSource() {

    }

    public static TasksRemoteDataSource getInstance() {
        if (INSTANCE == null) {
            synchronized (TasksRemoteDataSource.class) {
                if (INSTANCE == null) {
                    INSTANCE = new TasksRemoteDataSource();
                }
            }
        }
        return INSTANCE;
    }

    /**
     * {@link LoadTasksCallBack#onDataNotAvailable()} is never fired.
     * In a real remote data source implementation, this would be fired
     * if the server can't be contacted or the server returns an error.
     */
    @Override
    public void loadTasks(final LoadTasksCallBack callBack) {
        // Simulate network by delaying the execution.
        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                callBack.onTasksLoaded(new ArrayList<>(TASKS_SERVICE_DATA.values()));
            }
        }, SERVICE_LATENCY_IN_MILLIS);
    }

    /**
     * {@link GetTaskCallBack#onDataNotAvailable()} is never fired.
     * In a real remote data source implementation, this would be fired
     * if the server can't be contacted or the server returns an error.
     */
    @Override
    public void getTask(String taskId, final GetTaskCallBack callBack) {
        final TaskBean taskBean = TASKS_SERVICE_DATA.get(taskId);
        Handler handler = new Handler();
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                callBack.onTaskLoaded(taskBean);
            }
        }, SERVICE_LATENCY_IN_MILLIS);
    }

    @Override
    public void clearCompletedTasks() {
        Iterator<Map.Entry<String, TaskBean>> iterator = TASKS_SERVICE_DATA.entrySet().iterator();
        while (iterator.hasNext()) {
            Map.Entry<String, TaskBean> entry = iterator.next();
            if (entry.getValue().isCompleted()) {
                iterator.remove();
            }
        }
    }

    @Override
    public void refreshTasks() {

    }

    @Override
    public void addTask(TaskBean taskBean) {
        TASKS_SERVICE_DATA.put(taskBean.getId(), taskBean);
    }

    @Override
    public void updateTask(TaskBean taskBean) {
        TASKS_SERVICE_DATA.put(taskBean.getId(), taskBean);
    }

    @Override
    public void completeTask(TaskBean completedTaskBean) {
        completeTask(completedTaskBean.getId());
    }

    @Override
    public void completeTask(String taskId) {
        TaskBean taskBean = TASKS_SERVICE_DATA.get(taskId);
        if (taskBean != null) {
            taskBean.setCompleted(true);
            TASKS_SERVICE_DATA.put(taskBean.getId(), taskBean);
        }
    }

    @Override
    public void activateTask(TaskBean activeTaskBean) {
        activateTask(activeTaskBean.getId());
    }

    @Override
    public void activateTask(String taskId) {
        TaskBean taskBean = TASKS_SERVICE_DATA.get(taskId);
        if (taskBean != null) {
            taskBean.setCompleted(false);
            TASKS_SERVICE_DATA.put(taskBean.getId(), taskBean);
        }
    }

    @Override
    public void deleteTask(String taskId) {
        TASKS_SERVICE_DATA.remove(taskId);
    }

    @Override
    public void deleteAllTasks() {
        TASKS_SERVICE_DATA.clear();
    }

    private static void addTask(String title, String description) {
        TaskBean taskBean = new TaskBean(title, description, false);
        TASKS_SERVICE_DATA.put(taskBean.getId(), taskBean);
    }
}

File: AndroidArchitecture/mvvm-live-kotlin/src/main/java/com/codearms/maoqiqi/app/data/source/remote/TasksRemoteDataSource.kt
package com.codearms.maoqiqi.app.data.source.remote

import android.os.Handler
import com.codearms.maoqiqi.app.data.TaskBean
import com.codearms.maoqiqi.app.data.source.TasksDataSource
import java.util.*
import kotlin.collections.LinkedHashMap

/**
 * Access the server API interface data as a data source.
 * Author: fengqi.mao.march@gmail.com
 * Date: 2019/3/12 15:42
 */
class TasksRemoteDataSource private constructor() : TasksDataSource {

    /**
     * [TasksDataSource.LoadTasksCallBack.onDataNotAvailable] is never fired.
     * In a real remote data source implementation, this would be fired
     * if the server can't be contacted or the server returns an error.
     */
    override fun loadTasks(callBack: TasksDataSource.LoadTasksCallBack) {
        // Simulate network by delaying the execution.
        Handler().postDelayed({
            callBack.onTasksLoaded(ArrayList(TASKS_SERVICE_DATA.values))
        }, SERVICE_LATENCY_IN_MILLIS)
    }

    /**
     * [TasksDataSource.GetTaskCallBack.onDataNotAvailable] is never fired.
     * In a real remote data source implementation, this would be fired
     * if the server can't be contacted or the server returns an error.
     */
    override fun getTask(taskId: String, callBack: TasksDataSource.GetTaskCallBack) {
        val taskBean = TASKS_SERVICE_DATA[taskId]
        Handler().postDelayed({
            taskBean?.let { callBack.onTaskLoaded(it) }
        }, SERVICE_LATENCY_IN_MILLIS)
    }

    override fun clearCompletedTasks() {
        TASKS_SERVICE_DATA = TASKS_SERVICE_DATA.filterValues { it.isActive } as LinkedHashMap
    }

    override fun refreshTasks() {

    }

    override fun addTask(taskBean: TaskBean) {
        TASKS_SERVICE_DATA[taskBean.id] = taskBean
    }

    override fun updateTask(taskBean: TaskBean) {
        TASKS_SERVICE_DATA[taskBean.id] = taskBean
    }

    override fun completeTask(completedTaskBean: TaskBean) {
        completeTask(completedTaskBean.id)
    }

    override fun completeTask(taskId: String) {
        TASKS_SERVICE_DATA[taskId]?.let {
            it.isCompleted = true
            TASKS_SERVICE_DATA[it.id] = it
        }
    }

    override fun activateTask(activeTaskBean: TaskBean) {
        activateTask(activeTaskBean.id)
    }

    override fun activateTask(taskId: String) {
        TASKS_SERVICE_DATA[taskId]?.let {
            it.isCompleted = false
            TASKS_SERVICE_DATA[it.id] = it
        }
    }

    override fun deleteTask(taskId: String) {
        TASKS_SERVICE_DATA.remove(taskId)
    }

    override fun deleteAllTasks() {
        TASKS_SERVICE_DATA.clear()
    }

    companion object {
        private const val SERVICE_LATENCY_IN_MILLIS = 5000L

        @Volatile
        private var INSTANCE: TasksRemoteDataSource? = null

        private var TASKS_SERVICE_DATA: MutableMap<String, TaskBean> = LinkedHashMap(2)

        init {
            addTask("Build tower in Pisa", "Ground looks good, no foundation work required.")
            addTask("Finish bridge in Tacoma", "Found awesome girders at half the cost!")
        }

        @JvmStatic
        fun getInstance(): TasksRemoteDataSource = INSTANCE
                ?: synchronized(TasksRemoteDataSource::class.java) {
                    INSTANCE ?: TasksRemoteDataSource().also { INSTANCE = it }
                }

        private fun addTask(title: String, description: String) {
            val taskBean = TaskBean(title, description, false)
            TASKS_SERVICE_DATA[taskBean.id] = taskBean
        }
    }
}